Sources/SwiftSyntaxBuilder/Syntax+StringInterpolation.swift

```diff
@@ -192,3 +228,215 @@ extension SyntaxExpressibleByStringInterpolation {
     self.init(stringLiteral: source)
   }
 }
+
+// MARK: ExpressibleByLiteralSyntax conformances
+
+extension Collection {
+  fileprivate func allIndices(where predicate: @escaping (Element) -> Bool) -> UnfoldSequence<Index, Index> {
+    sequence(state: startIndex) { i in
+      guard let newI = self[i...].firstIndex(where: predicate) else {
+        return nil
+      }
+      i = index(after: newI)
+      return newI
+    }
+  }
+}
+
+extension Substring: ExpressibleByLiteralSyntax {
+  public func makeLiteralSyntax() -> ExprSyntaxProtocol {
+    // TODO: Choose whether to use a single-line or multi-line literal.
+    let quote = TokenSyntax.stringQuote
+
+    // Select a raw delimiter long enough that we won't need to escape quotes or backslashes.
+    // Locate backslashes and quotes...
+    let problemIndices = allIndices(where: #"\""#.contains(_:))
+    // Count adjacent hashes and compute the largest number (-1 = no problem chars)...
+    let maxPoundCount = problemIndices.reduce(-1) { prevCount, i in
```

Comment:
FWIW, we have an implementation like this over in `ConvenienceInitializers.swift`.

Reply:
I have some concerns about how that implementation handles non-interpolation backslashes in raw literals, but I suppose I should actually fix the thing if it's wrong instead of reimplementing it.

Reply:
See my comment above, it's not clear to me that we actually want/need that initializer any more.

Reply:
I've fixed up the convenience initializer and ripped out this separate implementation.